soa-derive-internal/src/iter.rs

```diff
@@ -15,243 +15,138 @@ pub fn derive(input: &Input) -> TokenStream {
-                                    .map(|field| &field.ty)
-                                    .collect::<Vec<_>>();
-    let first_field_type = &fields_types[0];
-
-    let mut iter_type = quote!{
-        slice::Iter<'a, #first_field_type>
-    };
-    let mut iter_pat = quote!{
-        #first_field
-    };
-    let mut create_iter = quote!{
-        self.#first_field.iter()
-    };
-
-    let mut iter_mut_type = quote!{
-        slice::IterMut<'a, #first_field_type>
-    };
-    let mut create_iter_mut = quote!{
-        self.#first_field.iter_mut()
-    };
-
-    if fields_types.len() > 1 {
-        for field in &input.fields[1..] {
-            let field_name = &field.ident;
-            let field_type = &field.ty;
-
-            iter_pat = quote!{
-                (#iter_pat, #field_name)
-            };
-
-            iter_type = quote!{
-                iter::Zip<#iter_type, slice::Iter<'a, #field_type>>
-            };
-
-            create_iter = quote!{
-                #create_iter.zip(self.#field_name.iter())
-            };
-
-            iter_mut_type = quote!{
-                iter::Zip<#iter_mut_type, slice::IterMut<'a, #field_type>>
-            };
-
-            create_iter_mut = quote!{
-                #create_iter_mut.zip(self.#field_name.iter_mut())
-            };
-        }
-    }
-
-    let mut generated = quote! {
+    let generated = quote!{
         #[allow(non_snake_case, dead_code)]
         mod #detail_mod {
             use super::*;
-            use std::slice;
-            #[allow(unused_imports)]
-            use std::iter;
 
-            #[allow(missing_debug_implementations)]
-            #visibility struct Iter<'a>(pub(super) #iter_type);
-
-            impl<'a> Iterator for Iter<'a> {
-                type Item = #ref_name<'a>;
-
-                #[inline]
-                fn next(&mut self) -> Option<#ref_name<'a>> {
-                    self.0.next().and_then(|#iter_pat|
-                        Some(#ref_name{
-                            #(#fields_names,)*
-                        })
-                    )
-                }
-
-                #[inline]
-                fn size_hint(&self) -> (usize, Option<usize>) {
-                    self.0.size_hint()
-                }
-            }
-
-            impl<'a> DoubleEndedIterator for Iter<'a> {
-
-                #[inline]
-                fn next_back(&mut self) -> Option<#ref_name<'a>> {
-                    self.0.next_back().and_then(|#iter_pat|
-                        Some(#ref_name{
-                            #(#fields_names,)*
-                        })
-                    )
-                }
-            }
-
-            impl #vec_name {
-                /// Get an iterator over the
-                #[doc = #ref_doc_url]
-                /// in this vector
-                #visibility fn iter(&self) -> Iter {
-                    Iter(#create_iter)
-                }
+            pub struct VecIter<'a> {
```

Comment:
Couldn't understand what the std was doing (`RawVec` doesn't have `.iter()`) because the code is not straight forward, so I made my simple iterator. Actually 2 days ago I realized that the standard is simply converting to slice and using its iterator. So yes, we could keep the old code with minor changes.

A point for a future debate would be if making a well-implemented custom iterator may be made more performant than using nested zips (but with the obvious downside of more code maintenance)